New Blog Post - Web SDK Observable Overhaul (#98)
vertigis/studio-dev-center: blog/2025-08-05-web-obervable-overhaul.mdx

Another thing we've noticed was a recurring misuse of `watch` from `reactiveUtils`. `watch` will return a handle that is supposed to be called to cleanup the watch when it is no longer needed. While this pattern of cleaning up a handler (or listener) is extremely prevalent in development, we’ve found that this is easily missed with `reactiveUtils` use, and have been known to miss it ourselves from time to time, causing memory leaks.

There is one more consideration that is easily missed when constructing watches using `reactiveUtils`, and to understand this, you need to know how JavaScript closures and lexical scoping work.

Let’s take the following code as an example:

```ts
function myLeakyFunc(
    someLargeObject: object,
    someObservableObject: object
) {
    // The lexical scope created at this level

>     // includes both `someLargeObject and `someObservableObject`.

Reviewer: ```suggestion
    // includes both `someLargeObject` and `someObservableObject`.
```

    // This lexical scope is captured by all 3 anonymous
    // functions' closures and will live for as long as the
    // anonymous functions are referenced.

    const myFunc = () => {
        console.log(someLargeObject);
    };

    const handle = watch(
        () => someObservableObject.prop,
        (newPropValue) => console.log(newPropValue)
    );

    return [myFunc, handle];
}
```

> On the surface this seems innocent enough. We have a large object that needs to be used in an anonymous function and a watch that needs to watch a property on a completely separate observable object and the anonymous function is returned for use by the caller as well as the watch handle for cleanup when no longer needed. But sadly this has the potential to create a whole host of memory leaks depending on the browser engine and garbage collector being used.

Reviewer: ```suggestion
On the surface this seems innocent enough. We have a large object that needs to be used in an anonymous function and a watch that needs to watch a property on a completely separate observable object, and the anonymous function is returned for use by the caller as well as the watch handle for cleanup when no longer needed. But sadly this has the potential to create a whole host of memory leaks depending on the browser engine and garbage collector being used.
```

While this last example is a basic one, it is easy to imagine how quickly these closure leaks can become a problem in modern JavaScript development.

### So, Back to the Removal of `.watch`

With the removal of `.watch` from the ArcGIS JS SDK classes it makes sense for us to remove our implementation of `.watch` as well, given that our `.watch` is tied into the ArcGIS JS SDK implementation. This would be a breaking change in the Web SDK which we do try to approach with much caution. While the removal of `.watch` from the ArcGIS JS SDK is going to be a breaking change for Web SDK developers, and there isn’t anything we can do about that, we always look carefully at any breaking changes we’re going to make to our APIs to ensure that they are truly in the best interests of progressing the ease of future development.

> Removing `.watch` would leave us relaying on `reactiveUtils` for observable functionality, and as you’ve seen reading through this, perhaps that isn’t the best approach for Web, with the use of the JS Proxy and various performance and memory issues. After much discussion, since we were going to have to make a breaking change to the Web SDK anyway with the removal of `.watch`, we decided to look into what it might look like if we revisited property observability with a fresh approach.

Reviewer: ```suggestion
Removing `.watch` would leave us relying on `reactiveUtils` for observable functionality, and as you’ve seen reading through this, perhaps that isn’t the best approach for Web, with the use of the JS Proxy and various performance and memory issues. After much discussion, since we were going to have to make a breaking change to the Web SDK anyway with the removal of `.watch`, we decided to look into what it might look like if we revisited property observability with a fresh approach.
```

## Shaping the New Watch System

### The Wish List

We compiled a wish list of everything we would want in a property observability implementation, knowing that we likely would need to compromise in some places.

1. Does not use a JS Proxy.
2. A watch system that would remain stable with the same interface if the ArcGIS JS SDK ever changes their watch system again.
3. Will error at compile time with TypeScript if given objects or properties that cannot be watched.
4. Will error or warn at runtime if trying to watch objects or properties that cannot be watched.
5. Will not leak memory, even when misused.

> 6. Will not do anything functional when nothing is watching a property that is watchable. IE: no tracking announcements, no change notifications to nowhere, etc.

Reviewer: ```suggestion
6. Will not do anything functional when nothing is watching a property that is watchable. i.e.: no tracking announcements, no change notifications to nowhere, etc.
```

The original `.watch` functions used a string to specify paths to properties, but it was just a string, causing mistypes and refactors to break watches without anything surfacing the issue in a timely manner. Those `.watch` functions were written quite awhile ago, when TypeScript wasn’t very good at recursive typing, and template string types didn’t exist.

So, along with the object to be watched we decided a string property path would be our best option. With modern TypeScript, we can now build out a complex, template string recursive type that can build out the available watch options for nested watch property paths. There is a bit of a performance hit in IDEs when this recursive type encounters large deeply nested objects, so after dialing in the type performance we had to limit the depth of recursion to three for full type checking. But we expect that with the new [native TypeScript](https://devblogs.microsoft.com/typescript/typescript-native-port/) compiler that’s in the works, this performance issue will get much better and we will likely be able to expand the depth of the type checking further.

> Aside from the minor IDE performance issue, there is another IDE related drawback to the string property path, you won’t be able to use rename functionality and have that update the string references nor will a reference lookup find watches for the properties being watched. But we decided these IDE drawbacks can be lived with, given the memory management advantage of this pattern.

Reviewer: ```suggestion
Aside from the minor IDE performance issue, there is another IDE related drawback to the string property path: you won’t be able to use rename functionality and have that update the string references, nor will a reference lookup find watches for the properties being watched. But we decided these IDE drawbacks can be lived with, given the memory management advantage of this pattern.
```

> After dialing in the types and starting to spread around the new watch, we did encounter one other quirk. When attempting to pass `this` as the object being watched, TypeScript wasn’t happy. Turns out, when you’re using constrained generic types for typing arguments and you pass `this` to one of those arguments TypeScript won’t infer the type of `this`. They have their reasons, but it is a bit inconvenient in our use case. We debated on how to handle this, and tried a number of different ways to workaround it, but in the end we couldn’t find anything better than just casting `this` when passed. Bt, we also realized that in a lot of cases we see where `this` is being passed as the watch object, a watch really shouldn’t be used. Using the watch system to know when a property changes that you have control of is pretty heavy handed; a better approach would be to do whatever action you want in a setter instead.

Reviewer: ```suggestion
After dialing in the types and starting to spread around the new watch, we did encounter one other quirk. When attempting to pass `this` as the object being watched, TypeScript wasn’t happy. Turns out, when you’re using constrained generic types for typing arguments and you pass `this` to one of those arguments TypeScript won’t infer the type of `this`. They have their reasons, but it is a bit inconvenient in our use case. We debated on how to handle this, and tried a number of different ways to workaround it, but in the end we couldn’t find anything better than just casting `this` when passed. But, we also realized that in a lot of cases we see where `this` is being passed as the watch object, a watch really shouldn’t be used. Using the watch system to know when a property changes that you have control of is pretty heavy handed; a better approach would be to do whatever action you want in a setter instead.
```

### No Base Class? Let’s try Decorators … (spoiler, we didn’t use decorators)

Wish list #9 would have us remove the `Observable` base class from all of our classes so that we can be more targeted about what is actually observable and what is not. To do this we would need some way to add functionality to a class without inheritance. This sounds like exactly what decorators are designed to do.
The first iteration of the new observable implementation was built as a class decorator. This would fulfill wish list item #9 and #11 (no base class and not specifying exact properties that are observable, respectively).

> However, we ran into some immediate problems around TypeScript typing. Mainly, TypeScript doesn’t augment the types of a class (or anything) that uses a decorator that, from the JavaScript perspective, does augment the class with new functionality. This created a typing problem, to satisfy wish list item #12 (opt in observability) we would need an extra interface that implementors would also have to add to their observable classes to ensure that the class is observable. It looked something like this:

Reviewer: ```suggestion
However, we ran into some immediate problems around TypeScript typing. Mainly, TypeScript doesn’t augment the types of a class (or anything) that uses a decorator that, from the JavaScript perspective, does augment the class with new functionality. This created a typing problem: to satisfy wish list item #12 (opt in observability) we would need an extra interface that implementors would also have to add to their observable classes to ensure that the class is observable. It looked something like this:
```

```ts
await whenOnce(obj, "myProp", { signal });
```

#### `createIHandle`

Creating a number of watches in one place is fairly common, which results in having a number of returned handles to deal with. To help simplify some situations we’ve added this utility function `createIHandle` that takes an `Iterable<IHandle>` and returns a single `IHandle`. This is useful when combined with `onWatch` but also other scenarios.

```ts
onWatch(this, "myProp", () => {
  const handles: IHandle[] = [];
  handles.push(
    watch(otherObj, "otherProp", () => notifyChange(this, "myProp");
    watch(anotherObj, "anotherProp", () => notifyChange(this, "myProp");
  );
  return createIHandle(handles);
});
```

The returned handle from `createIHandle` holds a reference to the iterable and will remove the handles currently held in the iterable when called, allowing you to update the iterable with new handles or remove old ones but still use the same returned handle to remove whatever is left in the iterable.

### What did we Deprecate?

> All classes that were inheriting from `Observable` (@vertigis/arcgis-extensions/support/Observable.ts) no longer inherit from `Observable`. Most have been updated to use the new mixin, however `SerializableBase` has not been updated to use the mixin as it did not make sense for this class to be observable on its own. Every class, including `SerializableBase`, that no longer inherit from `Observable` has had deprecated methods added to temporarily provide the previous methods from `Observable`. Support for `reactiveUtils` for all classes that were previously inheriting from `Observable` will still work as well during the transition period.

Reviewer: ```suggestion
All classes that were inheriting from `Observable` (@vertigis/arcgis-extensions/support/Observable.ts) no longer inherit from `Observable`. Most have been updated to use the new mixin, however `SerializableBase` has not been updated to use the mixin as it did not make sense for this class to be observable on its own. Every class, including `SerializableBase`, that no longer inherits from `Observable` has had deprecated methods added to temporarily provide the previous methods from `Observable`. Support for `reactiveUtils` for all classes that were previously inheriting from `Observable` will still work as well during the transition period.
```

> If you’re unsure if you have any references to anything that will be removed, you can check the browser console, all deprecated functions, classes, and modules that are still in use are logged there as warnings.

Reviewer: ```suggestion
If you’re unsure if you have any references to anything that will be removed, you can check the browser console; all deprecated functions, classes, and modules that are still in use are logged there as warnings.
```
